docs/RFCS/20191009_gc_protected_timestamps.md

```diff
@@ -0,0 +1,597 @@
+
+This RFC introduces a `protectedts` subsystem which provides primitives on top
+of which long running jobs can safely prevent relevant data from being garbage
+collected. This state is then plumbed in to the `storage` layer's data GC
+process to provide the specified semantics. We'll then detail how clients can
+use the transactional API to safely protect data needed for long-running jobs.
+
+## Basic terminology
+
+A protected timestamp is an HLC timestamp associated with a set of spans which,
+while it exists, prevents any range which contains data in those spans from
+garbage collecting data at that timestamp (open question about partial
+coverings, say Range contains [a, c) and [a, b) is protected, will [b,c) get
+collected?). There are two protection modes, `ProtectAt` and
+`ProtectAfter`. `ProtectAt` ensures only that values which are live at the
+specified timestamp are protected whereby `ProtectAfter` ensures that all
+values live at or after the timestamp are protected (note that `ProtectAfter`
+includes `ProtectAt` and protections marked as `ProtectAt` will be treated the
+sam in the initial implementation).  Each protected timestamp may contain
+metadata to identify the record (what exactly this metadata looks like is an
+open question).
+
+In this section we'll go through a high level overview of the basic operations
+and describe how they are made safe, describing the changes to the Replica's GC
+in the process, and then we'll talk about how we anticipate the API will be
+integrated.
+
+## Semantics of protecting a timestamp
+
+Clients of the protected timestamp subsystem would like it to be the case that
+if a timestamp is successfully protected then it is not possible that any data
+live in any of the protected timestamp's spans at the timestamp will be garbage
+collected until the protected timestamp is released.
+
+It would be unreasonable if an attempt to protect a timestamp succeeded after
+data live at that timestamp has already been GC'd. The protected timestamp
+subsystem should only succeed if indeed it is impossible for the timestamp
+which the client tries to protect is indeed protected. Furthermore it useful
+that clients be able to protect timestamps inside of a transaction rather than
+as a distinct transaction (i.e. the interface will take a `*client.Txn`).
+
+An important element to all of this is that GC TTLs are defined in 
+[Zone Configurations](https://www.cockroachlabs.com/docs/stable/configure-replication-zones.html).
+This implies a to need to couple the protected timestamp subsystem to the zone
+configuration subsystem or the downstream side-effects of those zone configs,
+replica state. This chooses to base its implementation on the zone config GC
+TTLs.
+
+There are different choices we could make about which timestamp are available
+for protection at any given time. This proposal allows clients to protect
+timestamps for a set of spans up to the GC ttl ago. Imagine that you have a 60
```

Comment:
From what you write above, it seems like you want the invariant that if the protection request comes back successfully, the data is actually protected, which means it has to actually touch all of the ranges containing any subset of the span to make sure the GC threshold isn't already violating what you're trying to protect (perhaps because the zone config's TTL was recently increased).